Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py

```diff
@@ -316,18 +316,30 @@ def pre_pr_checks(cls, repository):
             num_checks += 1
             name = key.split('.')[-1]
             log.info('    Running {}...'.format(name))
-            command = run(path.split(' '), cwd=repository.root_path)
-            if command.returncode:
-                if Terminal.choose(
+            for attempt in range(attempts):
+                command = run(path.split(' '), cwd=repository.root_path)
+                if command.returncode == 0:
+                    log.info('    Ran {}!'.format(name))
+                    break
+                options = ['Continue', 'Re-Run', 'No']
+                response = Terminal.choose(
                     '{} failed, continue uploading pull request?'.format(name),
+                    options=(options[0], options[2]) if attempt + 1 == attempts else options,
                     default='No',
-                ) == 'No':
+                )
+                if response == 'No':
                     sys.stderr.write('Pre-PR check {} failed\n'.format(name))
                     return False
-                else:
+                if response == 'Continue':
                     log.info('    {} failed, continuing PR upload anyway'.format(name))
-            else:
-                log.info('    Ran {}!'.format(name))
+                    break
+
+                if add_edits and run(
+                    [repository.executable(), 'commit', '-a', '--amend', '--date=now', '--no-edit'],
+                    cwd=repository.root_path,
+                ).returncode:
```

Comment:
I understand the use case, but committing during what is effectively a `git push` is way too magical for my liking. This change makes our workflow more like subversion and less like git, which IMO is the wrong direction to move in. 

With `-a`, I expect this to lead to unexpected changes being pushed (with the associated security/NPS concerns). It seems like it would also amend the wrong commit if you're pushing a multi-commit PR.

What I would vastly prefer is moving style checking to a pre-commit hook, where it's much less hassle to ask a developer to fix their style bug, re-add, and then restart the commit. I don't think we would need auto-rerun in pre-commit, but maybe there is an argument to have it for folks who prefer git-webkit to match webkit-patch as much as possible, and use `git-webkit pr` to make their commit.

Reply:
I think the big piece this comment misses is that many developers _want_ changes added when `git-webkit pr` is run (and while I'll let @smfr speak for himself, I believe the workflow here uses `git-webkit pr` without any commit and just modified files, so a pre-commit hooks doesn't really help).

I think your feedback is good, though, and it may be that the current design of this PR is overly-eager to add changes. In this iteration, adding files is powered by `--add` and `--no-add`, so it should do the same thing that `git-webkit pr` already does (although `-A` will add new files too, which is a problem and should be changed).

I think we should have a mode with `git-webkit` that does add style modifications, but I can see arguments that this should be configured by repository settings, would be interested to hear @smfr, @hortont424 and @marcoscaceres (who are involved in the motivating bug) what thing

Reply:
My primary request is to have the style checker run before `git webkit pr` makes the branch and commits. This feels more efficient than "let it run later and then re-run PR stuff". If someone has already committed and asks `git webkit pr` to not commit, then running it later seems fine.

Reply:
There is an interesting edge case, which is the reason this check is where it currently is: `check-webkit-style` won't check style of a new, but uncommitted, file. I think this means that we would basically have to run the style checker twice if we ran it pre-commit.

My understanding is that your primary concern, @smfr, is that you don't want to have to break out of `git-webkit` and add files manually after failing the style checker, right?

Reply:
+1 for @smfr comment above. 
Currently, if `git-webkit pr` finds any style error, the process to fix that and ensuring that those fixes become part of the pr being created is so non-intuitive/long.

Reply:
I agree, which was/is the point of this PR.

To be clear about what the behavior change proposed in this PR is: When a pre-pr check fails (at the moment, our pre-PR check is just the style checker, but I know some contributors make diff displaying a pre-PR check too), this PR will ask the user if they would like to continue uploading their PR, giving them 3 options:
- `Continue`: Ignore the failed check and upload the PR anyways
- `Re-Run`: If `git-webkit pr` is configured to automatically add and commit files, look for any changes and add them to the previous commit before re-running the failed check
- `No`: Exit the program.

My understanding of the frustration here is that the `Re-Run` option would basically fix them. In the event contributors prefer managing their own `git` state (like @emw-apple does), they have presumably either passed `--no-add` to `git-webkit pr` or configured `webkitscmpy.auto-create-commit` to be `false`, both options which disable the automatic commit amending in this PR.